app.ts

```diff
@@ -0,0 +1,298 @@
+    else if(from < 0 )
+      this._from = 0;
+    else
+      this._from = from;
+
+    this._to = this.to;
+
+  }
+
+  get to():number
+  {
+    let to:number;
+
+    to = this._from + this.getNumberOfRows()
+
+    if (to > this._totalRecords)
+      return this._totalRecords- 1;
+    else
+      return to;
+  }
+
+
+  //determine the amount of rows to add to the table based on the size on the window
+  getNumberOfRows()
+  { 
+    let rows:number;
+
+    //subtract - 1 to cater for header row.
+    let height = (parseInt(((window.innerHeight-75)/30).toFixed(0)) - 1);
+    let width = (parseInt(((window.innerWidth-75)/30).toFixed(0)) - 1);
+
+    if (height < width)
+      rows = height;
+    else
+      rows = width;
+
+    if (rows < 0)
+      return 0;
+    else
+      return rows;
+  
+  }
+
+  getTotalRecords()
+  {
+    const HttpRequest = new XMLHttpRequest();
+    const url='http://localhost:2050//recordCount';
+    HttpRequest.open("GET", url, false);
+    HttpRequest.send();
+
+    const responseText = HttpRequest.responseText
+    
+    return parseInt(responseText);
+  }
+
+  buildTable(from:number,to:number) 
+  { 
+    $("#dataTableBody").find("tr").remove();
+
+    let tableBody = <HTMLTableSectionElement> document.getElementById("dataTableBody");
+
+    console.log("test speed 1");
+
+    //make the ajax call to retrieve the records and build the table
+    $.ajax({
+      url: `http://localhost:2050/records?from=${from}&to=${to}`,
+      dataType:'json',
+      success:function(data) {
+
+        $.each(data, function(row)
+          {
+            const dataRow = tableBody.insertRow(-1);
+            $.each(data[row], function(cell)
+              {
+                const newCell = dataRow.insertCell(-1);
+                newCell.innerHTML = data[row][cell];
+              }
+            )
+          }
+        )
+      }
+    });
+
+    console.log("test speed 2");
+
+    tableBody = <HTMLTableSectionElement> document.getElementById("dataTableBody");
+
+    const dataRow = tableBody.insertRow(0);
+
+    //make the ajax call to retrieve the records and build the table header
+    $.ajax({
+      url: `http://localhost:2050/columns`,
+      dataType:'json',
+      async:false,
+      success:function(data) {
+  
+        //const dataRow = tableBody.insertRow(0);
+        $.each(data, function(cell)
+          {
+            const newCell = dataRow.insertCell(-1);
+            newCell.outerHTML = `<th>${data[cell]}</th>`;
+             
+          }
+        )
+      }
+    });
+
+    //call the button property set fucntion to set disable/enably buttons appropriately 
+    buttonPropertySet();
+  
+  }
+
+}
+
+class utility
+{
+  timeout:number = 0;
+
+  debounce(func: Function, wait:number){
+  
+    return (...args:any[]) => {
+      const later = () => {
+        clearTimeout(this.timeout);
+        func(...args);
+      };
+  
+      clearTimeout(this.timeout);
+      this.timeout = setTimeout(later, wait);
+    };
+  };
```

Comment:
I would prefer this function to look something like
```suggestion
      clearTimeout(this.timeout);
      this.timeout = setTimeout(() => {
          func(...args);
      }, wait);
    };
  };
```
This of course is once the functionality has been moved into the `table` (soon to be `Table`) class

This of course, being when it is not returning the callback, but rather executing the function on demand.